File: M06_Data_Projects/S02_Save_V03a_DB/app/src/main/java/com/example/save_v03a/MainActivity.java
package com.example.save_v03a;

import androidx.appcompat.app.AppCompatActivity;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.EditText;
import android.widget.TextView;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


/**
 * Using SQLite DB:
 *  => http://developer.android.com/training/basics/data-storage/databases.html
 *
 * Using AsyncTask:
 * => http://developer.android.com/guide/components/processes-and-threads.html
 *
 * This project demonstrates the structure of using the Android SQLite DB with an
 * AsyncTask.  Android does not permit long tasks on the UI thread, so AsyncTask is
 * used to perform DB actions in the background, and access UI thread in the
 * foreground for updating the UI.
 *
 */
public class MainActivity extends AppCompatActivity {

    EditText etResponse;
    TextView tvIsConnected;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // get reference to the views
        etResponse = (EditText) findViewById(R.id.etResponse);
        tvIsConnected = (TextView) findViewById(R.id.tvIsConnected);

        // check if you are connected or not
        tvIsConnected.setBackgroundColor(0xFF00CC00);
        tvIsConnected.setText("DB connected");

        // call to perform network operation on separate thread
        do_work_in_background();
        Log.d("Save_v03", "Main onCreate() complete");
    }

    // method does DB writing work in background
    private void do_work_in_background() {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        final Handler handler = new Handler(Looper.getMainLooper());

        executor.execute(new Runnable() {
            @Override
            public void run() {
                String result = "";

                //Background work here
                DBClass DBtest = new DBClass(MainActivity.this);

                // Add a row....
                {
                    // Gets the data repository in write mode
                    SQLiteDatabase db = DBtest.getWritableDatabase();

                    // Create a new map of values, where column names are the keys
                    ContentValues values = new ContentValues();
                    values.put("str_col", "TRUCK");
                    values.put("num_col", 600);

                    // Insert the new row, returning the primary key value of the new row
                    long newRowId;
                    newRowId = db.insert("sample_table", null, values);
                }

                //dump all rows
                {
                    SQLiteDatabase db = DBtest.getReadableDatabase();

                    // Define a projection that specifies which columns from the database
                    // you will actually use after this query.
                    String[] projection = {"id", "str_col", "num_col"};
                    String selection = "num_col < ?";  // ? gets filled in by args
                    String[] selectionArgs = {"850"};

                    // How you want the results sorted in the resulting Cursor
                    String sortOrder = "id" + " DESC";   // sort by descending id number

                    Cursor c = db.query(
                            "sample_table",  // The table to query
                            projection,                               // The columns to return
                            selection,                                // The columns for the WHERE clause
                            selectionArgs,                            // The values for the WHERE clause
                            null,                                     // don't group the rows
                            null,                                     // don't filter by row groups
                            sortOrder                                 // The sort order
                    );

                    c.moveToFirst();
                    long itemId = c.getLong(c.getColumnIndexOrThrow("id"));

                    Object o = null;  //object returned from column in DB
                    String key = "";
                    String value = "";

                    do {
                        int colCount = c.getColumnCount();
                        for (int i = 0; i < colCount; ++i) {
                            switch (c.getType(i)) {
                                case Cursor.FIELD_TYPE_INTEGER:
                                    key = c.getColumnName(i);
                                    value = String.valueOf(c.getInt(i));
                                    break;
                                case Cursor.FIELD_TYPE_STRING:
                                    key = c.getColumnName(i);
                                    value = c.getString(i);
                                    break;
                            }
                            Log.d("Save_v03", "key=" + key + " value=" + value);
                            result += "key=" + key + " value=" + value + "\n";
                        }
                        Log.d("Save_v03", "Next Row");
                        result += "\n";
                    } while (c.moveToNext());
                }

                Log.d("Save_v03", "Sleep ..........................");
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }


                // in case you want to show something on UI
                String finalResult = result;
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        //UI Thread work here
                        etResponse.setText("DONE!\n\n" + finalResult);
                    }
                });

                Log.d("Save_v03", "Background result..." + result);
            }
        });

    }

}

File: M05_Compose_Projects/C04_Compose_02_DB_Background/app/src/main/java/com/russ/compose_02/MainActivity.kt
package com.russ.compose_02

import android.content.ContentValues
import android.content.ContentValues.TAG
import android.database.Cursor
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.gestures.scrollable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import com.russ.compose_02.ui.theme.Compose_02Theme
import java.util.concurrent.Executors

// Kotlin Reference:  https://kotlinlang.org/docs/home.html
// Compose Reference: https://developer.android.com/jetpack/compose/documentation

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        do_work_in_background()  // Launch background DB Access

        // SetContent is compose call to assemble content ("compose" content)
        // Mouse-over setContent, Compose_01Theme, Surface, etc, to see
        // JavaDocs comments on what these are and what they do.
        setContent {
            // Compose_01Theme is declared in ui.theme
            Compose_02Theme {
                // A surface container using the 'background' color from the theme
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    // Now we run our own functions declared below, which will print
                    // to the screen using colour/surface/material per "Surface"
                    MainPage("Start .... wait for DB read ...")
                    Log.d(TAG, "onCreate() called")
                }
            }
        }
    }

    ///////////////////////////////// Access DB in Background //////////////////////////

    // method does DB writing work in background
    private fun do_work_in_background() {
        val executor = Executors.newSingleThreadExecutor()
        val handler = Handler(Looper.getMainLooper())
        executor.execute(object : Runnable {
            override fun run() {
                var result = ""

                //Background work here
                val DBtest = DBClass(this@MainActivity)

                // Add a row....
                run {

                    // Gets the data repository in write mode
                    val db = DBtest.writableDatabase

                    // Create a new map of values, where column names are the keys
                    val values = ContentValues()
                    values.put("str_col", "TRUCK")
                    values.put("num_col", 600)

                    // Insert the new row, returning the primary key value of the new row
                    val newRowId: Long
                    newRowId = db.insert("sample_table", null, values)
                }

                //dump all rows
                run {
                    val db = DBtest.readableDatabase

                    // Define a projection that specifies which columns from the database
                    // you will actually use after this query.
                    val projection =
                        arrayOf("id", "str_col", "num_col")
                    val selection = "num_col < ?" // ? gets filled in by args
                    val selectionArgs =
                        arrayOf("850")

                    // How you want the results sorted in the resulting Cursor
                    val sortOrder = "id" + " DESC" // sort by descending id number
                    val c = db.query(
                        "sample_table",  // The table to query
                        projection,  // The columns to return
                        selection,  // The columns for the WHERE clause
                        selectionArgs,  // The values for the WHERE clause
                        null,  // don't group the rows
                        null,  // don't filter by row groups
                        sortOrder // The sort order
                    )
                    c.moveToFirst()
                    val itemId = c.getLong(c.getColumnIndexOrThrow("id"))
                    val o: Any? = null //object returned from column in DB
                    var key = ""
                    var value = ""
                    while (c.moveToNext()) {
                        val colCount = c.columnCount
                        for (i in 0 until colCount) {
                            when (c.getType(i)) {
                                Cursor.FIELD_TYPE_INTEGER -> {
                                    key = c.getColumnName(i)
                                    value = c.getInt(i).toString()
                                }

                                Cursor.FIELD_TYPE_STRING -> {
                                    key = c.getColumnName(i)
                                    value = c.getString(i)
                                }
                            }
                            Log.d("Save_v03", "key=$key value=$value")
                            result += "key=$key value=$value\n"
                        }
                        Log.d("Save_v03", "Next Row")
                        result += "\n"
                    }
                }
                val d = Log.d("Save_v03", "Sleep ..........................")
                try {
                    Thread.sleep(5000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }


                // in case you want to show something on UI
                val finalResult = result
                handler.post { //UI Thread work here
                    // post result
                    setContent {
                        // Compose_01Theme is declared in ui.theme
                        Compose_02Theme {
                            // A surface container using the 'background' color from the theme
                            Surface(
                                modifier = Modifier.fillMaxSize(),
                                color = MaterialTheme.colorScheme.background
                            ) {
                                // Now we run our own functions declared below, which will print
                                // to the screen using colour/surface/material per "Surface"
                                MainPage(finalResult)
                                Log.d(TAG, "handler.post called")
                            }
                        }
                    }
                }
                Log.d("Save_v03", "Background result...$result")
            }
        })
    }


}

@Composable
fun MainPage(theData: String, modifier: Modifier = Modifier) {
    // Column allows putting text in a column (i.e. on the next row down)

    Column {
        // Add another column
        Text(text = "2nd line")

        // Function to add more columns
        Greeting2()

        var messages = arrayOf("Russ", "Was", "Here")
        Greeting3(messages)

        // Write the text to the screen
        Text(
            text = "$theData",
            modifier = Modifier.verticalScroll(rememberScrollState())
        )
    }
    Log.d(TAG, "Greeting() called")
}


/////////////// Add more lines
@Composable
fun Greeting2() {
    Column {
        Text(text = "Greeting2 - line 1 - ")
        Text(text = "Greeting2 - line 2")
        Text(text = "Greeting2 - line 3")
    }
    Log.d(TAG, "Greeting() called")
}

////////////// Add from an Array
@Composable
fun Greeting3(things: Array<String>) {
    Column {
        Text(text = "  --------- Start -----------")
        for (thing in things) {
            Text(thing)
        }
        Text(text = "  --------- Done -----------")
    }
    Log.d(TAG, "Greeting() called")
}

